src/Stan/Config.hs

```diff
@@ -96,3 +103,81 @@ finaliseConfig :: PartialConfig -> Trial Text Config
+  where
+    filesMap :: HashMap FilePath (HashSet (Id Inspection))
+    filesMap = HashMap.fromList $ map (, inspectionsIds) paths
+
+    useCheck
+        :: HashMap FilePath (HashSet (Id Inspection))
+        -> Check
+        -> HashMap FilePath (HashSet (Id Inspection))
+    useCheck dict Check{..} =
+        applyForScope (applyFilter checkType checkFilter) checkScope dict
+
+    applyFilter
+        :: CheckType
+        -> Maybe CheckFilter
+        -> HashSet (Id Inspection)
+        -> HashSet (Id Inspection)
+    applyFilter = \case
+        Include -> includeFilter
+        Ignore -> ignoreFilter
+
+    ignoreFilter :: Maybe CheckFilter -> HashSet (Id Inspection) -> HashSet (Id Inspection)
+    ignoreFilter cFilter = HashSet.filter (not . satisfiesFilter cFilter)
+
+    includeFilter :: Maybe CheckFilter -> HashSet (Id Inspection) -> HashSet (Id Inspection)
+    includeFilter mFilter ins = case mFilter of
+         -- add all inspections to the existing ones
+         Nothing -> inspectionsIds <> ins
+         Just cFilter -> case cFilter of
+             CheckInspection iId -> HashSet.insert iId ins
+             CheckObservation _ -> ins
+             CheckSeverity sev ->
+                 let sevInspections = filter ((== sev) . inspectionSeverity) inspections
+                 in HashSet.fromList (map inspectionId sevInspections) <> ins
+             CheckCategory cat ->
+                 let catInspections = filter (elem cat . inspectionCategory) inspections
+                 in HashSet.fromList (map inspectionId catInspections) <> ins
+
+    -- Returns 'True' if the given inspection satisfies 'CheckFilter'
+    satisfiesFilter :: Maybe CheckFilter -> Id Inspection -> Bool
+    satisfiesFilter mFilter iId = case mFilter of
+        Nothing -> True  -- no filter, always satisfies
+        -- TODO: rewrite more efficiently after using GHC-8.10
+        Just cFilter -> case lookupInspectionById iId of
+            Nothing -> False  -- no such ID => doesn't satisfy
+            Just Inspection{..} -> case cFilter of
+                CheckInspection checkId -> iId == checkId
+                CheckObservation _      -> False
+                CheckSeverity sev       -> sev == inspectionSeverity
+                CheckCategory cat       -> elem cat inspectionCategory
+
+    applyForScope
+        :: (HashSet (Id Inspection) -> HashSet (Id Inspection))
+        -> Maybe CheckScope
+        -> HashMap FilePath (HashSet (Id Inspection))
+        -> HashMap FilePath (HashSet (Id Inspection))
+    applyForScope f mScope hm = case mScope of
+        Nothing -> f <$> hm  -- no scope = apply for everything
+        Just cScope -> case cScope of
+            CheckScopeFile path -> HashMap.alter (fmap f) path hm
+            CheckScopeDirectory dir ->
+                HashMap.fromList
+                $ mapKeepMaybe
+                    (\(path, ins) -> guard (isInDir dir path) *> pure (path, f ins))
+                $ HashMap.toList hm
```

Comment:
Maybe instead of converting into list here, we can somehow use `HashMap.mapWithKey` in here?